spec/java/util/stream/Stream.main.lsl

```diff
@@ -0,0 +1,1049 @@
+
+        var i: int = 0;
+        action LOOP_FOR(
+            i, 0, this.length, +1,
+            _accumulate_with_biConsumer_loop(i, accumulator, result)
+        );
+
+    }
+
+
+    @Phantom proc _accumulate_with_biConsumer_loop (i: int, accumulator: BiConsumer, result: Object): void
+    {
+        action CALL(accumulator, [result, this.storage[i]]);
+    }
+
+
+    fun *.collect (@target self: Stream, collector: Collector): Object
+    {
+        if (collector == null)
+            _throwNPE();
+
+        var i: int = 0;
+        val accumulator: BiConsumer = action CALL_METHOD(collector, "accumulator", []);
+        action LOOP_FOR(
+            i, 0, this.length, +1,
+            _accumulate_with_biConsumer_loop(i, collector, result)
+        );
+    }
+
+
+    fun *.min (@target self: Stream, comparator: Comparator): Optional
+    {
+        if (comparator == null)
+            _throwNPE();
+
+        if (this.length == 0)
+        {
+            result = action DEBUG_DO("Optional.empty()");
+        }
+        else
+        {
+            var min: Object = this.storage[0];
+
+            var i: int = 0;
+            action LOOP_FOR(
+                i, 1, this.length, +1,
+                _find_min_loop(i, comparator, min)
+            );
+
+            result = action DEBUG_DO("Optional.of(min)");
+        }
+    }
+
+
+    @Phantom proc _find_min_loop (i: int, comparator: Comparator, min: int): void
+    {
+        if (action CALL(comparator, [min, this.storage[i]]) > 0)
+            min = this.storage[i];
+    }
+
+
+    fun *.max (@target self: Stream, comparator: Comparator): Optional
+    {
+        if (comparator == null)
+            _throwNPE();
+
+        if (this.length == 0)
+        {
+            result = action DEBUG_DO("Optional.empty()");
+        }
+        else
+        {
+            var max: Object = this.storage[0];
+
+            var i: int = 0;
+            action LOOP_FOR(
+                i, 1, this.length, +1,
+                _find_max_loop(i, comparator, max)
+            );
+
+            result = action DEBUG_DO("Optional.of(max)");
+        }
+    }
+
+
+    @Phantom proc _find_max_loop (i: int, comparator: Comparator, max: int): void
+    {
+        if (action CALL(comparator, [max, this.storage[i]]) < 0)
+            max = this.storage[i];
+    }
+
+
+    fun *.count (@target self: Stream): long
+    {
+        result = this.length;
+    }
+
+
+    fun *.anyMatch (@target self: Stream, predicate: Predicate): boolean
+    {
+        if (predicate == null)
+            _throwNPE();
+
+        result = false;
+
+        var i: int = 0;
+        action LOOP_FOR(
+            i, 0, this.length, +1,
+            _anyMatch_loop(i, predicate, result)
+        );
+    }
+
+
+    @Phantom proc _anyMatch_loop (i: int, predicate: Predicate, result: boolean): void
+    {
+        if (action CALL(predicate, [this.storage[i]]))
+        {
+            result = true;
+            action LOOP_BREAK();
```

Comment:
A `LOOP_WHILE` can be used here instead of calling `LOOP_BREAK`.

Reply:
Fixed

Reply:
Please avoid using the same `@Phantom proc` as loop body for different instances of `LOOP_*` actions. This makes it confusing to read and follow through the control logic and data flow.

I suggest the following style:
```dart
fun foo (...): ...
{
    action LOOP_WHILE(
        result && a < b,
        foo_loop(a, b, result)   // if there are only a single loop in this function/subroutine
        // foo_something (...)   // if there is multiple loops, named with distinct sounding name
        // foo_loop_outer (...)  // if there is nested loops
    );
}

@Phantom proc foo_loop (...): void
{
...
}
```